### ISSUE TITLE
Implement formatter similar to JSCS default reporter

### ISSUE THREAD

**Issue by bdukes** [eslint/eslint#5860]
The default JSCS reporter includes the context of the code and an indicator to the specific character which is causing the violation.  With JSCS joining ESLint, that's the one JSCS feature I'd miss (and love to see added to ESLint).

As an example, it can look something like this:

    All identifiers must be camelCase or UPPER_CASE at C:\inetpub\wwwroot\site.local\JavaScript\BaseLocator.js :
        66 |        this._noLocationSpan = value;
        67 |    },
        68 |    get_instructionSpan: function() {
    ------------^
        69 |        return this._instructionSpan;
        70 |    },

**Comment by hzoo** [comment]
Yep that's definetely something I wanted to get over

**Comment by IanVS** [comment]
There is already https://github.com/royriojas/eslint-friendly-formatter, which at least prints the line and shows the position.  Is that mostly what you're looking for?

**Comment by nzakas** [comment]
Yeah, we should add this and should be fairly straightforward to do. I think the biggest question here is what would be the right name for the reporter? "JSCS" makes sense right now, but in another year or two, would that be confusing?

**Comment by nzakas** [comment]
This would also probably be a great first issue for one of the JSCS folks. :)

**Comment by lydell** [comment]
Perhaps it makes sense to use [babel-code-frame](https://github.com/babel/babel/tree/master/packages/babel-code-frame), just like [AVA added a few days ago](https://github.com/sindresorhus/ava/pull/742#issuecomment-208174649).

**Comment by nzakas** [comment]
@lydell while I appreciate you promoting your creation, I think we'll probably just port what's in JSCS. ;-)

**Comment by hzoo** [comment]
@nzakas Just to be clear, @lydell didn't create babel-code-frame (it was always part of Babel) - he was just the last person to make a PR to it (to fix the tab spacing issue when AVA used it).

**Comment by nzakas** [comment]
Thanks for the clarification @hzoo AMD apologies @lydell for misrepresenting your comment. 

I'd still like to port what's in JSCS. @hzoo does the default JSCS formatter have a name other than "default"?

**Comment by ikokostya** [comment]
Current reporter in `jscs` prints only 2 lines before and after. I propose to add option for this.

**Comment by nzakas** [comment]
@ikokostya we don't have options for formatters, but you can always create your own formatter that does exactly what you want.

**Comment by ikokostya** [comment]
@nzakas  But if [babel-code-frame](https://github.com/babel/babel/tree/master/packages/babel-code-frame) will be a part of `eslint`, how I can configure it (e.g. set [highlightCode](https://github.com/babel/babel/tree/master/packages/babel-code-frame#options) option)?

**Comment by nzakas** [comment]
@ikokostya I think you misunderstood the discussion here. Someone just suggested that `babel-code-frame

**Comment by nzakas** [comment]
@ikokostya I think you misunderstood the discussion. Someone suggested using `babel-code-frame` to implement this formatter, but that does not mean you would have direct access to `babel-code-frame` or its options. In reality, we probably won't use `babel-code-frame` at all and will just port what is in JSCS already.

Formatters in ESLint do not accept options. We provide some formatters out of the box, and you can also create your own.

If you have further questions about this, please stop by the chatroom: https://gitter.im/eslint/eslint

**Comment by vitorbal** [comment]
I opened an issue where I propose passing the "context" lines of code around a problem reported by ESLint, which would enable us to fully port the JSCS formatter over. Would love to hear your feedback on the proposal: https://github.com/eslint/eslint/issues/7098

**Comment by hzoo** [comment]
Is there a particular reason for not using code-frame? https://github.com/babel/babel/blob/master/packages/babel-code-frame/README.md#usage - the api is pretty simple and it acts the same as JSCS. I added an option for changing the lines above/below as well.

**Comment by vitorbal** [comment]
If we don't mind the extra dependency, then I don't see why not. I tried it out and it works quite well:

<img width="907" alt="screenshot 2016-09-29 12 12 36" src="https://cloud.githubusercontent.com/assets/626038/18950767/dbcbe73e-8641-11e6-99f7-00a82aca8b35.png">

The pointer renders a bit differently from the JSCS formatter though:
```bash
  1 | class Foo {
> 2 |   constructor()
    |                ^
  3 | }
```
vs.:
```bash
  1 | class Foo {
  2 |   constructor()
---------------------^
  3 | }
```

**Comment by lydell** [comment]
@vitorbal If you try the latest version, the colors are nicer too :)

**Comment by hzoo** [comment]
Yep!

<img width="622" alt="screen shot 2016-09-29 at 8 27 26 am" src="https://cloud.githubusercontent.com/assets/588473/18953891/92e3e152-861e-11e6-9915-2e98526b7cc4.png">

If we want the `---` we can make that an option too.

I removed the babel-runtime dep so it's just https://github.com/babel/babel/blob/master/packages/babel-code-frame/package.json#L11-L13

    "chalk": "^1.1.0", // used in eslint already, for colors
    "esutils": "^2.0.2", // used in eslint already
    "js-tokens": "^2.0.0" // just regex to differentiate tokens

**Comment by vitorbal** [comment]
Oh, nice! Thanks for the screenshots. Mine were taken before the newest version.
This looks great, I'd be 👍 for using it. Let's see what the other folks think.

**Comment by vitorbal** [comment]
I have a question about the formatter output: should we replicate the JSCS formatter 1 to 1, or could I take the liberty of making a few simple improvements? For example, here are two:

1. With JSCS, the formatter would output just the filename next to the error message:
<img width="829" alt="screenshot 2016-10-01 00 59 38" src="https://cloud.githubusercontent.com/assets/626038/19009278/5a667f9c-8772-11e6-855f-553bd8d00e38.png">
However, I think there would be more value in outputting the entire filepath, like this:
<img width="830" alt="screenshot 2016-10-01 01 01 05" src="https://cloud.githubusercontent.com/assets/626038/19009290/8defbb44-8772-11e6-82b3-f3ffef37ff0c.png">
2. JSCS doesn't differentiate errors vs. warnings, but ESLint does, so it would be helpful if each message could display if itself is an error or a warning. Compare:
<img width="569" alt="screenshot 2016-10-01 01 10 24" src="https://cloud.githubusercontent.com/assets/626038/19009434/386c424e-8774-11e6-8edb-6c419979c8dd.png">  
vs.:
<img width="886" alt="screenshot 2016-10-01 01 18 56" src="https://cloud.githubusercontent.com/assets/626038/19009514/10097726-8775-11e6-927c-4ade55dd1c17.png">

Any opinions?

**Comment by ilyavolodin** [comment]
@vitorbal 👍 on proposed enhancements

**Comment by nzakas** [comment]
Agreed. These seem like good enhancements

**Comment by zxqfox** [comment]
Would be great to have line and column in the filename link to use cmd+click behaviour

**Comment by a-x-** [comment]
e.g.:
```
tmp2.js:1:1: Do not use 'new' for side effects. [Error/no-new]
tmp2.js:1:5: 'FooBar' is not defined. [Error/no-undef]
tmp2.js:1:11: Missing semicolon. [Error/semi]
```

* relative paths (to project root)
* :line:column like eslint `unix` formatter
* colored output like jscs and eslint default formatter (stylish)

**Comment by vitorbal** [comment]
@zxqfox Great idea! That's actually the last "feature" I am working on for the JSCS formatter before I can open the PR for it :)
